### PR TITLE
Fixes for Spring core integration

### PR DESCRIPTION
**Pull request by gregw**
Misc fixes to improve the spring integration with jetty core:
 + Improved list API for HttpFields (still not entirely consistent)
 + Fixed a bug in the ContentSinkSubscriber that was causing write pending exceptions

### CODE REVIEW

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java [hunk 1]
```diff
@@ -668,4 +668,47 @@ public long getLongValue()
             return _long;
         }
     }
+
+    public static class MultiHttpField extends HttpField
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java]: Make it `private` as we don't want people start using this class if then we change the implementation.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java [hunk 2]
```diff
@@ -989,6 +990,31 @@ default Mutable add(HttpFields fields)
             return this;
         }
 
+        /**
+         * Add a field with a list value;
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java]:
```suggestion
         * <p>Adds a field associated with a list of values.</p>
```

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java [hunk 3]
```diff
@@ -1186,27 +1212,24 @@ default Mutable put(HttpHeader header, String value)
         }
 
         /**
-         * Set a field.
+         * Set a field to a list value;
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java]:
```suggestion
         * <p>Puts a field associated with a list of values.</p>
```

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpGeneratorClientTest.java [hunk 4]
```diff
@@ -98,7 +98,7 @@ public void testEmptyHeaders() throws Exception
 
         HttpFields.Mutable fields = HttpFields.build();
         fields.add("Host", "something");
-        assertThrows(IllegalArgumentException.class, () -> fields.add("Null", null));
+        assertThrows(IllegalArgumentException.class, () -> fields.add("Null", (String)null));
```
Comment (sbordet) [inline jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpGeneratorClientTest.java:102]: I fear this is a breaking change, as now existing code that passes `null` has to add a cast and recompile.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java [hunk 5]
```diff
@@ -989,6 +990,31 @@ default Mutable add(HttpFields fields)
...
         }
 
+        /**
+         * Add a field with a list value;
+         *
+         * @param name the name of the field
+         * @param list the List value of the field. If null the field is cleared.
+         * @return this builder
+         */
+        default Mutable add(String name, List<String> list)
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:1000]: If this is going to clash with `add(String, String)` when passing `null`, how about renaming it `addMulti()` to avoid the signature ambiguity?

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkSubscriber.java [hunk 6]
```diff
@@ -76,6 +87,8 @@ public void onError(Throwable failure)
     @Override
     public void onComplete()
     {
-        sink.write(true, null, callback);
+        // wait until called twice: once from last write success and once from the publisher
+        if (lastAndComplete.decrementAndGet() == 0)
+            callback.succeeded();
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkSubscriber.java]: This method cannot be called twice, as it violates the spec, for example if a class subclasses `ContentSinkSubscriber`.

Can you detail what is the problem that lead to this fix?
Seems a wrong fix if the publisher of chunks respects reactive streams.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkSubscriber.java (hunk 6 above)
Comment (sbordet) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkSubscriber.java]: I pushed a fix after checking the reactive streams specification, and added a clarifying comment.

**Comment by sbordet** [comment]
@gregw let me know if you want me to take over this PR.
Overall just small changes.
For reactive streams, and I double checked the specification and pushed fixes.

**Comment by gregw** [comment]
@sbordet I'm working most afternoons, so no need for you to take over this PR... unless release is happening tomorrow.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java (hunk 5 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:1000]: @sbordet  we'd need to rename the put(String name, List<String> list) as well...
So since we had the put version, I'm almost thinking this is OK... who would add explicit null?  Currently that would throw an IAE, so the code would not work anyway.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java]: best I can do is package protected.

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpGeneratorClientTest.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpGeneratorClientTest.java:102]: Existing code that passes null to add will throw IAE, so I think this is OK.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java (hunk 5 above)
Comment (sbordet) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:1000]: @gregw then the javadoc needs to be changed, as it says that `null` clears the field.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java [hunk 7]
```diff
@@ -989,6 +990,33 @@ default Mutable add(HttpFields fields)
...
+                return this;
+            if (list.size() == 1)
+            {
+                String v = list.get(0);
+                if (v == null)
+                    return this;
+                return add(name, v);
+            }
+            HttpField field = new HttpField.MultiHttpField(name, list);
+            if (field.getValueList() == null)
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java]: This line is strange, as we know the list has more than 1 element at this point.
How can be given to the field, and the field return `null`?
At least it deserves a comment, e.g. "In case all values are blanks".

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java [hunk 8]
```diff
@@ -1186,27 +1214,24 @@ default Mutable put(HttpHeader header, String value)
...
-            }
-            return this;
+            Objects.requireNonNull(name);
+            if (list == null || list.isEmpty())
+                return remove(name);
+            if (list.size() == 1)
+                return put(name, list.get(0));
+
+            HttpField field = new HttpField.MultiHttpField(name, list);
+            if (field.getValueList() == null)
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java]: Ditto.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java [hunk 9]
```diff
@@ -668,4 +668,47 @@ public long getLongValue()
...
+                if (StringUtil.isBlank(v))
+                    continue;
+                if (builder == null)
+                    builder = new StringBuilder(list.size() * v.length() * 2);
+                else
+                    builder.append(", ");
+                builder.append(v);
+            }
+
+            return builder == null ? null : builder.toString();
```
Comment (sbordet) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java:696]: In case of `List.of(" ", " ")`, we return `null`, but since fields with an empty value are allowed, should we instead return the empty string?

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java [hunk 10]
```diff
@@ -1000,13 +1000,61 @@ public void testPutNullName()
         assertThat(fields.size(), is(0));
     }
 
+    @Test
+    public void testAddNullValueList()
```
Comment (sbordet) [inline jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java:1004]: Add test for empty list, and list with more than 1 blank string.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java:696]: I've been struggling with this.  If we allow blanks, then should the result be " , " or "" or null?
I think the simplest thing to do is to turn null entries into "" and then allow blank list entries as as well.  It means there can be list values like ", ,   ,   , " , which are invalid according to https://www.rfc-editor.org/rfc/rfc9110#section-5.6.1, but then we don't prevent invalid values from being set directly.   Also that suggests that blank list entries are OK, so long as there is at least one non blank, which is harder to enforce and explain.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java (hunk 7 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java]: removed (see below)

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java (hunk 8 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java]: ditto

**Comment by gregw** [comment]
@lachlan-roberts @joakime can I get a review of this, as @sbordet is away.

**Review by joakime — DISMISSED** [review]

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java [hunk 11]
```diff
@@ -1000,13 +1000,90 @@ public void testPutNullName()
...
+        list.add(null);
+        fields.add("name", list);
+        assertThat(fields.size(), is(1));
+        assertThat(fields.get("name"), is(""));
+
+        list.add("");
+        fields.clear();
+        fields.add("name", list);
+        assertThat(fields.size(), is(1));
+        assertThat(fields.get("name"), is(", "));
```
Comment (sbordet) [inline jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java]: Such value is explicitly forbidden by the spec.

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java [hunk 12]
```diff
@@ -1000,13 +1000,90 @@ public void testPutNullName()
...
+
+        list.add(" ");
+        fields.clear();
+        fields.add("name", list);
+        assertThat(fields.size(), is(1));
+        assertThat(fields.get("name"), is(", ,  "));
```
Comment (sbordet) [inline jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java]: Ditto.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java (hunk 9 above)
Comment (sbordet) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java:696]: Ok I am confused.

I think we should discard `null` and blank strings, as the spec in 5.6.1.1 says that the sender must not generate empty elements.
But the tests explicitly test for values such as ", ,  ", which I don't think it's what we want.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java:696]: @sbordet Now I'm confused?  You said above that you wanted to convert the invalid list `List.of(" ", " ")` into the valid `""` empty value.  I could not find a sensible way of doing that generally.   If we just skip null/blank items, we can end up with the empty list, then it is unclear if that is a remove or an add of the empty value.

So instead I took the approach of what we do with other `add`/`put` methods, in that we do not enforce validity.   I think I explained that well enough above.

I think not enforcing validity is reasonable.   I could also live with throwing an IAE if we are given and invalid list.  I'm not so keen on trying to convert an invalid list into something valid.

File: jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java (hunk 11 above)
Comment (gregw) [reply jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java]: I know.   But you said you wanted to accept such lists in your previous review???

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java:696]: I have changed to throw IAE.

**Review by sbordet — APPROVED** [review]
